src/ledger/LedgerManagerImpl.cpp

```diff
@@ -2791,4 +2835,299 @@ LedgerManagerImpl::ApplyState::addAnyContractsToModuleCache(
+        for (auto const& tx : originalSorobanTxs)
+        {
+            // Create a new transaction frame to avoid modifying the original
+            newSorobanTxs.emplace_back(
+                TransactionFrame::makeTransactionFromWire(mApp.getNetworkID(),
+                                                          tx->getEnvelope()));
+        }
+
+        // Build parallel tx set
+        auto sorobanConfig =
+            mApp.getLedgerManager().getSorobanNetworkConfigForApply();
+        sorobanConfig.setLedgerMaxDependentTxClusters(
+            std::atoi(std::getenv("STELLAR_TEST_PARALLEL_EXECUTION")));
+        auto laneConfig =
+            createSurgePricingLaneConfig(TxSetPhase::SOROBAN, mApp, true);
+        std::vector<bool> hadTxNotFittingLane;
+
+        // Use the current ledger version for building stages
+        uint32_t ledgerVersion =
+            std::max((uint32_t)PARALLEL_SOROBAN_PHASE_PROTOCOL_VERSION,
+                     (uint32_t)Config::CURRENT_LEDGER_PROTOCOL_VERSION);
+        auto parallelStages = buildSurgePricedParallelSorobanPhase(
+            newSorobanTxs, mApp.getConfig(), sorobanConfig, laneConfig,
+            hadTxNotFittingLane, ledgerVersion);
+
+        if (parallelStages.empty())
+        {
+            CLOG_WARNING(Ledger,
+                         "Failed to build parallel stages for re-execution");
+            return;
+        }
+
+        CLOG_INFO(Ledger, "Built {} parallel stages for testing",
+                  parallelStages.size());
+
+        // Find the starting index for Soroban transactions in results
+        uint32_t sorobanStartIndex = 0;
+        for (size_t i = 0; i < sorobanPhaseIndex; ++i)
+        {
+            sorobanStartIndex += originalPhases[i].sizeTx();
+        }
+
+        // Create mutable results for parallel execution
+        std::vector<MutableTxResultPtr> parallelMutableResults;
+        parallelMutableResults.reserve(newSorobanTxs.size());
+
+        // Map to track original tx to its index
+        std::unordered_map<Hash, size_t> txHashToParallelIndex;
+
+        // Get Soroban network config for fee calculations
+        auto const& sorobanNetworkConfig =
+            mApp.getLedgerManager().getSorobanNetworkConfigForApply();
+
+        for (size_t i = 0; i < newSorobanTxs.size(); ++i)
+        {
+            auto const& tx = newSorobanTxs[i];
+            txHashToParallelIndex[tx->getContentsHash()] = i;
+
+            // Create completely fresh result
+            auto mutableResult = tx->createValidationSuccessResult();
+
+            // Initialize refundable fee tracker for Soroban transactions
+            // This is normally done in processFeeSeqNum/commonPreApply
+            if (tx->isSoroban())
+            {
+                // Cast to TransactionFrame to access
+                // computePreApplySorobanResourceFee
+                auto const& txFrame = static_cast<TransactionFrame const&>(*tx);
```

Comment:
This cast is not valid because you might be dealing with a fee bump tx here. We should probably expose this functionality via proper virtual functions for tests.

Reply:
fixed